packages/contracts-bedrock/test/L2/L2StandardBridgeInterop.t.sol

```diff
@@ -0,0 +1,361 @@
+
+// Target contract dependencies
+import {
+    L2StandardBridgeInterop,
+    InvalidDecimals,
+    InvalidLegacyAddress,
+    InvalidSuperchainAddress,
+    InvalidTokenPair,
+    IOptimismMintableERC20Factory,
+    MintableAndBurnable
+} from "src/L2/L2StandardBridgeInterop.sol";
+import { IERC20Metadata } from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+import { IERC165 } from "@openzeppelin/contracts/utils/introspection/IERC165.sol";
+import { IOptimismMintableERC20 } from "src/universal/IOptimismMintableERC20.sol";
+
+// TODO: Replace Predeploys.OPTIMISM_SUPERCHAIN_ERC20_FACTORY with optimismSuperchainERC20Factory
+import { Predeploys } from "src/libraries/Predeploys.sol";
+
+contract L2StandardBridgeInterop_Test is Bridge_Initializer {
+    /// @notice Emitted when a conversion is made.
+    event Converted(address indexed from, address indexed to, address indexed caller, uint256 amount);
+
+    /// @notice Mock ERC20 decimals
+    function _mockDecimals(address _token, uint8 _decimals) internal {
+        vm.mockCall(_token, abi.encodeWithSelector(IERC20Metadata.decimals.selector), abi.encode(_decimals));
+    }
+
+    /// @notice Mock ERC165 interface
+    function _mockInterface(address _token, bytes4 _interfaceId, bool _supported) internal {
+        vm.mockCall(
+            _token, abi.encodeWithSelector(IERC165.supportsInterface.selector, _interfaceId), abi.encode(_supported)
+        );
+    }
+
+    /// @notice Mock factory deployment
+    function _mockDeployments(address _factory, address _token, address _deployed) internal {
+        vm.mockCall(
+            _factory,
+            abi.encodeWithSelector(IOptimismMintableERC20Factory.deployments.selector, _token),
+            abi.encode(_deployed)
+        );
+    }
+}
+
+/// @notice Test when converting from a legacy token to a SuperchainERC20 token
+contract L2StandardBridgeInterop_LegacyToSuper_Test is L2StandardBridgeInterop_Test {
+    function _setUpLegacyToSuper(address _from, address _to) internal {
+        // Assume
+        vm.assume(_from != console2.CONSOLE_ADDRESS);
+        vm.assume(_to != console2.CONSOLE_ADDRESS);
+
+        // Mock same decimals
+        _mockDecimals(_from, 18);
+        _mockDecimals(_to, 18);
+
+        // Mock `_from` to be a legacy address
+        _mockInterface(_from, type(IERC165).interfaceId, true);
+        _mockInterface(_from, type(IOptimismMintableERC20).interfaceId, true);
+    }
+
+    /// @notice Test that the `convert` function with different decimals reverts
+    function testFuzz_convert_differenteDecimals_reverts(
+        address _from,
+        uint8 _decimalsFrom,
+        address _to,
+        uint8 _decimalsTo,
+        uint256 _amount
+    )
+        public
+    {
+        // Assume
+        vm.assume(_from != console2.CONSOLE_ADDRESS);
+        vm.assume(_to != console2.CONSOLE_ADDRESS);
+        vm.assume(_decimalsFrom != _decimalsTo);
+        vm.assume(_from != _to);
```

Comment:
If I remove this check, it will pass either way. That's not being checked on the contract.
Should we add it on the contract? `_from` and `_to` should never have the same address, right? Or it can happen

Reply:
You can try to use it with the same address and will fail in the factories checker because of only existing in one of them

Reply:
ah good then!